### ISSUE TITLE
Locally installed ipython3 not working with elpy on Pixelbook

### ISSUE THREAD

**Issue by vnckppl** [jorgenschaefer/elpy#1737]
# Summary
I installed ipython with 
```
pip3 install ipython --user
```
Elpy throws an error when I start a python shell. ipython works fine from the terminal. Emacs shows me the following errors when I try to invoke an ipython shell from a buffer:

*Warnings*

> Warning (python): Python shell prompts cannot be detected.
> If your emacs session hangs when starting python shells
> recover with ‘keyboard-quit’ and then try fixing the
> interactive flag for your interpreter by adjusting the
> ‘python-shell-interpreter-interactive-arg’ or add regexps
> matching shell prompts in the directory-local friendly vars:
>   + ‘python-shell-prompt-regexp’
>   + ‘python-shell-prompt-block-regexp’
>   + ‘python-shell-prompt-output-regexp’
> Or alternatively in:
>   + ‘python-shell-prompt-input-regexps’
>   + ‘python-shell-prompt-output-regexps’
> Warning (python): Python shell prompts cannot be detected.
> If your emacs session hangs when starting python shells
> recover with ‘keyboard-quit’ and then try fixing the
> interactive flag for your interpreter by adjusting the
> ‘python-shell-interpreter-interactive-arg’ or add regexps
> matching shell prompts in the directory-local friendly vars:
>   + ‘python-shell-prompt-regexp’
>   + ‘python-shell-prompt-block-regexp’
>   + ‘python-shell-prompt-output-regexp’
> Or alternatively in:
>   + ‘python-shell-prompt-input-regexps’
>   + ‘python-shell-prompt-output-regexps’
> Warning (python): Python shell prompts cannot be detected.
> If your emacs session hangs when starting python shells
> recover with ‘keyboard-quit’ and then try fixing the
> interactive flag for your interpreter by adjusting the
> ‘python-shell-interpreter-interactive-arg’ or add regexps
> matching shell prompts in the directory-local friendly vars:
>   + ‘python-shell-prompt-regexp’
>   + ‘python-shell-prompt-block-regexp’
>   + ‘python-shell-prompt-output-regexp’
> Or alternatively in:
>   + ‘python-shell-prompt-input-regexps’
>   + ‘python-shell-prompt-output-regexps’
> 

*Python*

> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> import argparse
> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> import argparse
> Traceback (most recent call last):
>   File "/home/vk/.local/bin/ipython3", line 5, in <module>
>     from IPython import start_ipython
> ModuleNotFoundError: No module named 'IPython'
> 
> Process Python exited abnormally with code 1
> 

# My configuration
## OS
Pixelbook Version 80.0.3987.18 (Official Build) dev (64-bit)

## Result of `(elpy-config)`
```
Emacs.............: 26.3
Elpy..............: 1.32.0
Virtualenv........: None
Interactive Python: /home/vk/.local/bin/ipython3 . (/home/vk/.local/bin/ipython3)
RPC virtualenv....: rpc-venv (/home/vk/.emacs.d/elpy/rpc-venv)
 Python...........: python3 3.7.4 (/usr/bin/python3)
 Jedi.............: Not found (0.15.2 available)
 Rope.............: Not found (0.14.0 available)
 Autopep8.........: Not found (1.4.4 available)
 Yapf.............: Not found (0.29.0 available)
 Black............: Not found (19.10b0 available)
Syntax checker....: Not found (flake8)
```

## Elpy configuration in my init.el
```
;;;; elpy
(elpy-enable)

(setenv "IPY_TEST_SIMPLE_PROMPT" "1")
(setq python-shell-interpreter "/home/vk/.local/bin/ipython3"
      python-shell-interpreter-args "-i")

(when (require 'flycheck nil t)
  (setq elpy-modules (delq 'elpy-module-flymake elpy-modules))
  (add-hook 'elpy-mode-hook 'flycheck-mode))

```

**Comment by galaunay** [comment]
The [doc](https://elpy.readthedocs.io/en/latest/ide.html#interpreter-setup) recommends to use the argument `--simple-prompt` as well:
```elisp
(setq python-shell-interpreter "/home/vk/.local/bin/ipython3"
      python-shell-interpreter-args "-i --simple-prompt")
```

Is it enough to fix your problem ?

**Comment by vnckppl** [comment]
Thank you for your reply. 
Unfortunately, this did not resolve the error.

**Comment by vnckppl** [comment]
I have completely removed all python and iPython versions from my system. There is a python3 (3.5.3) installation that comes with the pixelbook crostini container which is installed under:
`/usr/bin/python3`

I reinstalled iPython via:
`pip3 install ipython`

iPython3 is installed under:
`/usr/local/bin/ipython3`
iPython starts with python 3.5.3

In my .emacs, I have set the following elpy configuration:
```
(elpy-enable)
(setq elpy-rpc-python-command "/usr/local/bin/ipython3")

(setenv "IPY_TEST_SIMPLE_PROMPT" "1")
(setq python-shell-interpreter "/usr/local/bin/ipython3"
      python-shell-interpreter-args "-i --simple-prompt")

```

elpy-config shows:

> Elpy Configuration
> 
> Emacs.............: 26.3
> Elpy..............: 1.32.0
> Virtualenv........: None
> Interactive Python: /usr/local/bin/ipython3 (not found)
> RPC virtualenv....: rpc-venv (/home/vk/.emacs.d/elpy/rpc-venv)
>  Python...........: python3 3.7.4 (/usr/bin/python3)
>  Jedi.............: Not found (0.15.2 available)
>  Rope.............: Not found (0.14.0 available)
>  Autopep8.........: Not found (1.4.4 available)
>  Yapf.............: Not found (0.29.0 available)
>  Black............: Not found (19.10b0 available)
> Syntax checker....: Not found (flake8)

Not surprisingly, when I start a python shell in emacs, I receive the following error:

> ‘/usr/local/bin/ipython3’ cannot be found. Please set ‘python-shell-interpreter’ to a valid python binary.

Interestingly, when I set my elpy to python3 instead of ipython3:

```
(setq python-shell-interpreter "/usr/bin/python3"
      python-shell-interpreter-args "-i")
```

emacs is able to open a python shell, but opens python3 v3.7.4
also, this python version does not have access to the modules that I have installed on my machine which I can access with my python 3.5.3 version. 

Elpy config with python3 looks like:

>Elpy Configuration
> 
> Emacs.............: 26.3
> Elpy..............: 1.32.0
> Virtualenv........: None
> Interactive Python: /usr/bin/python3 3.7.4 (/usr/bin/python3)
> RPC virtualenv....: rpc-venv (/home/vk/.emacs.d/elpy/rpc-venv)
>  Python...........: python3 3.7.4 (/usr/bin/python3)
>  Jedi.............: Not found (0.15.2 available)
>  Rope.............: Not found (0.14.0 available)
>  Autopep8.........: Not found (1.4.4 available)
>  Yapf.............: Not found (0.29.0 available)
>  Black............: Not found (19.10b0 available)
> Syntax checker....: Not found (flake8)

How is it possible that my system and elpy when pointed to the same binary find different versions of python (with access to different libraries), and that elpy can't find the ipython binary, even though it exists in my path?

**Comment by galaunay** [comment]
This confuse me as well as I can't see how Elpy could do such thing...
There is still a few things you can try to get a bit more insight on this.

### About the RPC
> (setq elpy-rpc-python-command "/usr/local/bin/ipython3")

the RPC should be a non-interactive python, setting this back to `"/usr/bin/python3"` or similar and running `M-x elpy-rpc-reinstall-virtualenv`, should fix some of your issues.

### About the executable not found
> Interactive Python: /usr/local/bin/ipython3 (not found)

I extracted the code that Elpy uses to find the python executable:
``` elisp
(defun elpy-display-python-version ()
  (interactive)
  (let* ((interactive-python (if (boundp 'python-python-command)
                                 python-python-command
                               python-shell-interpreter))
         (exec (executable-find interactive-python))
         (version (let ((pversion (shell-command-to-string
                                   (format "%s --version"
                                           python-shell-interpreter))))
                    (when (string-match "[0-9.]+" pversion)
                      (match-string 0 pversion)))))
    (message "interactive python: %s" interactive-python)
    (message "python executable: %s" exec)
    (message "python executable version: %s" version)))
```
This function will just display the found executable and its version.
Could you try executing it with different values of `python-shell-interpreter` and report what is displayed in the `*Messages*` buffer ?

**Comment by vnckppl** [comment]
Thank you so much for your help.

I made the suggested changes to my .emacs, but that did not fix the issue.
Then I ran the code to get the python version for each setting for `python-shell-interpreter` which returned python 3 or 3.5.3 depending on the input. 

I decided to reinstall emacs. I had previously installed emacs 26 with flatpak, which as it turned out caused all these problems. After removing flatpak emacs and reinstalling emacs (25) via apt-get, everything is working again.
